### PR TITLE
SELV3-659: Add period mapping between OLMIS and DHIS2

### PR DESCRIPTION
**Pull request by sradziszewski**
Jira task link: https://openlmis.atlassian.net/browse/SELV3-659

### CODE REVIEW

File: src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  private LocalDate endDate;
+
+  @ManyToOne(fetch = FetchType.LAZY)
+  @JoinColumn(name = "serverId", nullable = false)
+  private Server server;
+
+  /**
+   * Creates new instance based on data from the importer.
+   */
+  public static PeriodMapping newInstance(Importer importer) {
```
Comment (kmadej) [inline src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java:79]: it should be done via constructor, not static method

File: src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    exporter.setSource(source);
+    exporter.setDhisPeriod(dhisPeriod);
+    exporter.setProcessingPeriodId(processingPeriodId);
+    exporter.setStartDate(startDate);
+    exporter.setEndDate(endDate);
+    exporter.setServer(server);
+  }
+
+  public interface Exporter extends BaseExporter {
+
```
Comment (kmadej) [inline src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java:115]: remove empty line

File: src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    void setDhisPeriod(String dhisPeriod);
+
+    void setProcessingPeriodId(UUID processingPeriodId);
+
+    void setStartDate(LocalDate startDate);
+
+    void setEndDate(LocalDate endDate);
+
+    void setServer(Server server);
+
```
Comment (kmadej) [inline src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java:129]: remove empty line

File: src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  }
+
+  public interface Importer extends BaseImporter {
+
```
Comment (kmadej) [inline src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java:133]: remove empty line

File: src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    String getDhisPeriod();
+
+    UUID getProcessingPeriodId();
+
+    LocalDate getStartDate();
+
+    LocalDate getEndDate();
+
+    Server getServer();
+
```
Comment (kmadej) [inline src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java:147]: remove empty line

File: src/main/java/org/openlmis/integration/dhis2/dto/periodmapping/PeriodMappingDto.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+  @JsonFormat(shape = STRING)
+  private LocalDate endDate;
+
+  private ServerDto serverDto;
+
+  /**
+   * Creates new instance based on domain object.
+   */
+  public static PeriodMappingDto newInstance(PeriodMapping periodMapping) {
```
Comment (kmadej) [inline src/main/java/org/openlmis/integration/dhis2/dto/periodmapping/PeriodMappingDto.java:60]: it should be done via construtor

File: src/main/java/org/openlmis/integration/dhis2/web/period/PeriodMappingController.java [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    return PeriodMappingDto.newInstance(periodMappingToSave);
+  }
+
+  /**
+   * Deletes the specified period mapping.
+   */
+  @DeleteMapping(value = "/{id}")
+  @ResponseStatus(HttpStatus.NO_CONTENT)
+  public void deletePeriodMapping(@PathVariable("id") UUID id) {
+    if (!periodMappingRepository.existsById(id)) {
```
Comment (kmadej) [inline src/main/java/org/openlmis/integration/dhis2/web/period/PeriodMappingController.java:157]: i do not see any logic to delete period mapping

**Review by kmadej — CHANGES_REQUESTED** [review]

File: src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java (hunk 1 above)
Comment (sradziszewski) [reply src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java:79]: static factory method was used due to the presence of such a convention in other parts of the system, e.g. 
[referencedata](https://github.com/OpenLMIS/openlmis-referencedata/blob/master/src/main/java/org/openlmis/referencedata/domain/Orderable.java#L291), [dhis2-integration](https://github.com/villagereach/selv-v3-dhis2-integration/blob/master/src/main/java/org/openlmis/integration/dhis2/domain/dataset/Dataset.java#L87)

File: src/main/java/org/openlmis/integration/dhis2/dto/periodmapping/PeriodMappingDto.java (hunk 6 above)
Comment (sradziszewski) [reply src/main/java/org/openlmis/integration/dhis2/dto/periodmapping/PeriodMappingDto.java:60]: static factory method was used due to the presence of such a convention in other parts of the system, e.g.
[referencedata](https://github.com/OpenLMIS/openlmis-referencedata/blob/master/src/main/java/org/openlmis/referencedata/domain/Orderable.java#L291), [dhis2-integration](https://github.com/villagereach/selv-v3-dhis2-integration/blob/master/src/main/java/org/openlmis/integration/dhis2/domain/dataset/Dataset.java#L87)

File: src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java (hunk 2 above)
Comment (sradziszewski) [reply src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java]: Line removed

File: src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java (hunk 3 above)
Comment (sradziszewski) [reply src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java]: Line removed

File: src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java (hunk 4 above)
Comment (sradziszewski) [reply src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java]: Line removed

File: src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java (hunk 5 above)
Comment (sradziszewski) [reply src/main/java/org/openlmis/integration/dhis2/domain/periodmapping/PeriodMapping.java]: Line removed

File: src/main/java/org/openlmis/integration/dhis2/web/period/PeriodMappingController.java (hunk 7 above)
Comment (sradziszewski) [reply src/main/java/org/openlmis/integration/dhis2/web/period/PeriodMappingController.java:157]: added deleting PeriodMapping logic

**Review by kmadej — APPROVED** [review]
looks good to me

**Pull request merged**